My Jetpack: Add Backup 'needs-attention' red bubble and notice banner

elliottprogrammer opened the pull request:
This PR add a red bubble and notice/banner when the backup 'needs-attention' status is triggered (when a backup failure is detected).

<!--- Provide a general summary of your changes in the Title above -->
<!-- Would you like this feature to be tested by Beta testers?
Please add testing instructions to projects/plugins/jetpack/to-test.md in a new commit as part of your PR. -->
<!-- a12s: If you have an expected version that you're aiming for the PR to add, please use the Milestone field to communicate it. If you leave it blank, that indicates there isn't a preference. -->
Screenshot: 

<img width="1306" alt="Screen Shot 2024-12-11 at 08 20 58" src="https://github.com/user-attachments/assets/c32772bb-973b-4639-9ee1-7884b94f834c">


Fixes: https://github.com/Automattic/jetpack-roadmap/issues/1430

## Proposed changes:
<!--- Explain what functional changes your PR includes -->
* Adds a red bubble when a BAckup failure is detected.
* Displays a red bubble notice/banner (Top of My Jetpack page) with message, primary CTA the says "Read troubleshooting guide" and secondary CTA that says "Contact support".

### Other information:

- [ ] Have you written new tests for your changes, if applicable?
- [ ] Have you checked the E2E test CI results, and verified that your changes do not break them?
- [ ] Have you tested your changes on WordPress.com, if applicable (if so, you'll see a generated comment below with a script to run)?

## Jetpack product discussion
<!-- If you're an Automattician, include a shortlink to the p2 discussion with Jetpack Product here. -->
<!-- Make sure any changes to existing products have been discussed and agreed upon -->

Project Thread: pbNhbs-bJN-p2
Design post: p1HpG7-umC-p2

## Does this pull request change what data or activity we track or use?
<!--- If so, please add the "[Status] Needs Privacy Updates" label and explain what changes there are. -->
<!--- Check existing Jetpack support documents for a preview of the information we need. -->
No

## Testing instructions:
<!-- If you were reviewing this PR, how would you like the instructions to be presented? -->
<!-- Please include detailed testing steps, explaining how to test your change. -->
<!-- Bear in mind that context you working on is not obvious for everyone.  -->
<!-- Adding "simple" configuration steps will help reviewers to get to your PR as quickly as possible. -->
<!-- "Before / After" screenshots can also be very helpful when the change is visual. -->

**Testing notes:** To test this PR, basically we need to force a backup to fail. I tried various ways to make a backup fail, but most my attempts were unsuccessful. The only way I've been able to make a failed backup is by making the site unaccessible, and then queueing the backup to run from the MC Rewind Debugger. To make the site unaccessible, I simply killed my Jurassic Tube tunneling, but there are other various ways to do this such as by stopping Apache, blocking port 80, inserting `die()` or `exit()` command in index.php, etc..
There is another promising way of making a backup fail due to a "No such file" error, but I believe the method only works on a self-hosted site that is Not running within a Docker container. (I was not able to get it to work on my local dev environment or with Jurassic Ninja (confirmed it does not work on a JN site here: p1733501055164569/1733349201.389539-slack-CS8UYNPEE )
Instructions on how to carry out this method are here: _Generating “No such file” errors during backup_: pbuNQi-1Wi-p2
If you know of any other ways to trigger a backup failure, please let me know!

### Testing steps:

* Run this PR branch locally in a [local Jetpack dev environment](https://github.com/Automattic/jetpack/blob/trunk/docs/development-environment.md#docker-supported-recommended) (via Docker & JurassicTube or Ngrok tunneling).
* Once Jetpack is up and running, ensure it is fully connected. (Also ensure you are NOT sandboxed, and NOT using STORE_SANDOX or anything like that.)
* Purchase a Backup plan (using A11N credits).
* Wait for your first full backup to complete.
* Kill your Jurassic Tube or Ngrok tunneling, so the site is no longer accessible via the tunneling domain.
* Open your site in the MC Rewind Debugger: PCYsg-lNo-p2
* Queue up a new full backup on your site.
* Wait until the full backup completes:
    - in the MC Rewind debugger page you can find the current backup in the "Recent Full Backup Attempts" section and wait for it to show that is done (it should finish with the status 'not-accessible').
    - You can also open the WPCOM API Developer Console: https://developer.wordpress.com/docs/api/console/. (sign in)
        - Check the rewind backups endpoint at `wpcom/v2`, endpoint: `/sites/239657125/rewind/backups?force=wpcom`, and verify the status of the last backup is `not-available`.
* Now re-connect your Jurassic Tube or Ngrok tunneling.
* Open your site ad go to My Jetpack page.
* At the top of My Jetpack you should see a Red bubble and notice banner: (See screenshot)

![Screen Shot 2024-12-18 at 12 17 54](https://github.com/user-attachments/assets/11e14fad-1f0a-4c3c-8885-17eff80ab225)


* Confirm the button to "Read troubleshooting guide" points to: https://jetpack.com/support/backup/troubleshooting-jetpack-backup/
* Confirm the "Contact support" button points to: https://jetpack.com/support/
* Ensure the Notice/banner looks good in all viewport sizes & mobile sizes.
* Look at the Backup product card. The status of the backup card should be "Needs attention". The card should have a red border around it, and the CTA should say, "Troubleshoot" which points to the Jetpack.com Troubleshooting backup page.
* Also the Backup card should say, "The last backup attempt failed." with a tooltip. And also say "Check out our troubleshooting guide" (see screenshot:)
* Click to open the tooltip. Verify it looks like the screenshots below:

![Screen Shot on 2024-12-18 at 12-15-52](https://github.com/user-attachments/assets/5c127a65-bb78-437b-9544-df13d93b0446)

![Screen Shot on 2024-12-18 at 12-16-12](https://github.com/user-attachments/assets/36aea825-ed26-491e-8fcb-89244dd6edbc)

* Next, disconnect Jetpack. The red bubble and notice should now be the "Missing site connection" notice. Verify there are no console errors.
* Now establish a site-only connection.  In this case there should now be 2 red bubbles. 1 for missing user connection, and 1 for Backup needs-attention, but the missing user connection notice is higher priority so it should be showing. Ensure there are no console errors.
* Now re-establish user connection, and you should now see the Backup needs-attention notice again.

Note: In a followup PR I think I should implement allowing the user to optionally close(x) each red bubble notice (persistent for a temporary amount of time).

elliottprogrammer committed:
Add red bubble alert if Backups failing.

elliottprogrammer committed:
changelog

elliottprogrammer committed:
Add 'needs-attention' status when Backup fails.

elliottprogrammer committed:
Remove Red bubble alert code, for use in followup PR.

elliottprogrammer committed:
Clean up, remove testing/dev code.

elliottprogrammer committed:
Fix doctype return value.

elliottprogrammer committed:
Phan suppress PhanTypeSuspiciousNonTraversableForeach warning.

elliottprogrammer committed:
Update changelog entry - incorrect description.

elliottprogrammer committed:
Add redbubble & notice when backups are failing.

elliottprogrammer committed:
Call Products\Backup class directly. fix phan TypeError.

elliottprogrammer committed:
Add changelog.

elliottprogrammer committed:
Adjust the red bubble notice priorities (reverse order).

CodeyGuyDylan reviewed (changes requested):
When testing this, I noticed that the Backup day in the body of the notice is a day ahead of my local day (i.e., it tested this on Dec 9th, but it's saying Dec 10th was the last time it tried to backup the site)
![image](https://github.com/user-attachments/assets/fd8b2a00-7087-4408-9135-68f146cb7a71)

I assume it's going off UTC time, how hard would it be to localize this? I updated my site's timezone just to be safe, but the day was still wrong.

Also, I don't know if we've had a notice like this before with this much content, the mobile styling needs to be addressed because this looks pretty bad 😅 
![image](https://github.com/user-attachments/assets/5e30c80d-cc8d-41aa-9810-140d608f6870)

(I think I designed this so that's on me)

Other than those things, this tested correctly and I think the code looks great 😄

CodeyGuyDylan commented:
> Note: In a followup PR I think I should implement allowing the user to optionally close(x) each red bubble notice (persistent for a temporary amount of time). What do you think? Any thoughts?

I think that would be a great idea! As a user, I hate non-dismissible notices

elliottprogrammer committed:
Fix Notice styling for mobile view.

elliottprogrammer committed:
Fix last Backup date: Apply local timezone.

elliottprogrammer committed:
changelog

elliottprogrammer commented:
> When testing this, I noticed that the Backup day in the body of the notice is a day ahead of my local day (i.e., it tested this on Dec 9th, but it's saying Dec 10th was the last time it tried to backup the site) 
> 
> I assume it's going off UTC time, how hard would it be to localize this? I updated my site's timezone just to be safe, but the day was still wrong.
> 
> Also, I don't know if we've had a notice like this before with this much content, the mobile styling needs to be addressed because this looks pretty bad 😅 
> 
> (I think I designed this so that's on me)
> 
> Other than those things, this tested correctly and I think the code looks great 😄

Thanks for reviewing @CodeyGuyDylan! 🙌
- Fixed the backup time discrepancy between UTC & local time (in the Notice).  Now displaying the backup day **_and time_** in the Notice.
- Fixed mobile styling in the Notice/banner.
- Added some additional context in the Backup product card:
    - Now says, "The last backup attempt failed.", and shows the error code.
- Added `self::STATUS_NEEDS_ATTENTION` to the `$warning_statuses` array.
- I updated the PR description/testing instructions and screenshots to reflect the above changes.

elliottprogrammer committed:
Add backup failed message in product card.

elliottprogrammer committed:
Merge branch 'trunk' into add/backup-needs-attention-status

elliottprogrammer committed:
Merge branch 'add/backup-needs-attention-status' into add/backup-redbubble-needs-attention-notice

CodeyGuyDylan commented:
> > > Note: In a followup PR I think I should implement allowing the user to optionally close(x) each red bubble notice (persistent for a temporary amount of time). What do you think? Any thoughts?
> > 
> > 
> > I think that would be a great idea! As a user, I hate non-dismissible notices
> 
> I'm just realizing that in my comment, I make it sound like this is **All my idea**, lol! 🤦‍♂️. This is unintentional and I want to clarify that @CodeyGuyDylan made this suggestion (allowing the notice to close) in a [previous related PR](https://github.com/Automattic/jetpack/pull/39816#pullrequestreview-2459330878) (Displaying expired plan notices), and as I'm developing more of these notices, I'm agreeing more and more that it should be implemented. 😊

I did not remember suggesting this 😆

CodeyGuyDylan reviewed (changes requested):
I'm still seeing some design issues on the specific screen widths of 481px-585px, below is the worst case scenario:
![image](https://github.com/user-attachments/assets/13b0309e-174a-4a6e-a2c4-3a0a135472fc)

I think this looks alright, we may want to confirm with design before shipping though if you haven't already, they may have some tweaks 😄 
![image](https://github.com/user-attachments/assets/e3050523-6c64-47a3-9a09-92691326022b)

Note: I think we could potentially replace `Save every change` with the error in this case 🤔 it looks a bit jumbled with both there

elliottprogrammer committed:
Moar fixes to Notice mobile view.

elliottprogrammer commented:
Hey @keoshi! 👋,
Design review requested:
[From feedback it was suggested](https://github.com/Automattic/jetpack/pull/40454#pullrequestreview-2489869756) that we add some sort of content within the Backup product card stating **the reason why** Backup "Needs attention" & troubleshooting.
So from that, I added an error icon with text saying, "The last backup attempt failed". And below that I added the "Error code: [the error code/message returned from the server]"
Can you take a look and provide any feedback or suggestions? Thanks! 🙂

keoshi commented:
Thanks for the ping @elliottprogrammer, appreciate it!

> I added an error icon with text saying, "The last backup attempt failed".

Love this and it more closely connects the card to the message above. Thanks for adding this!

> And below that I added the "Error code: [the error code/message returned from the server]"

This one I have some reservations about, just because it makes it look more technical than it should be. If we can indeed point to a specific error — `not-accessible` in this case —, couldn't we just make that part of the message above, like so:

> Backup failed: Your site was inaccessible during the last attempt.

This way we lead with the problem and then explain why that happened. This will clean up the card and, by consequence, make the troubleshoot button stand out that much more.

Also understand this could be a lot more work with mapping different sentences for every possible error state the API might throw, so feel free to push back on what I said above. :)

Thanks for being such good shepherds of great UX, @elliottprogrammer and @CodeyGuyDylan !

elliottprogrammer committed:
Improve detection of failed backup.

elliottprogrammer commented:
**Update:** I'm currently working to try to map all the possible Backup/Rewind error codes, and translate them into human readable descriptions...  So we can at least give the user a basic idea/reason why the backup attempt failed.  Standby...

keoshi commented:
@elliottprogrammer Thanks for doing that!

> That would take additional time to reach out to the Backup team to ask them what all the possible error codes are, and what they mean, and IMO I thinking it's out of scope and not worth the time and effort.

Agreed this could either be tackled later or, anything that's super vague and/or doesn't make sense to expose, maybe we can just show a generic error?

> there are various other error codes (Rewind errors) that either I don't know what they mean, or are too vague, such as: unknown_error, vp_can_transfer, host_not_supported, vp_active_on_site, and there are other various individual Backup error codes that come from the VaultPress API (which I don't have access to) so I don't even know what the error codes are.

In parallel with the above, I wonder if there's anyone of your team @jwebbdev that would be able to support Bryan. We could even create a FG page document these errors, what they mean, and what the human equivalent copy should be? Thanks in advance!

elliottprogrammer committed:
Add 'source' property to needs-attention status data.

elliottprogrammer commented:
I don't seem to be getting a response to the question above, so I asked this question in the #yamato-backups channel in Slack, here: p1734441559123969-slack-CS8UYNPEE

elliottprogrammer committed:
Merge branch 'trunk' into add/backup-needs-attention-status

elliottprogrammer committed:
Merge branch 'add/backup-needs-attention-status' into add/backup-redbubble-needs-attention-notice

elliottprogrammer committed:
Add human readable Backup error descriptions to product card and Notice.

elliottprogrammer commented:
@CodeyGuyDylan ,
I pinged the Backup team and they provided me with a better understanding and descriptions of all the possible backup error codes. - here: p1734441559123969-slack-CS8UYNPEE

I've updated the code, to display more human readable descriptions (when a Backup failure occurs) within a tooltip in the Backup product card, as well as in the red bubble Notice/banner. (See screenshots:)

![Screen Shot on 2024-12-18 at 12-15-52](https://github.com/user-attachments/assets/3da264be-594c-4522-8f08-6cc2a127ad58)

![Screen Shot on 2024-12-18 at 12-16-12](https://github.com/user-attachments/assets/bd4bc309-8597-4aec-8d78-823918d5a98f)

![Screen Shot 2024-12-18 at 12 17 54](https://github.com/user-attachments/assets/55d82115-3a74-48b0-82e1-14dde6d7d8f1)

I've updated the PR testing instructions & screenshots. 👍

CodeyGuyDylan commented on projects/packages/my-jetpack/_inc/components/product-cards-section/backup-card/index.jsx:
```diff
@@ -128,12 +130,58 @@ const getTimeSinceLastRenewableEvent = lastRewindableEventTime => {
+							</InfoTooltip>
+						</Text>
+						<Text variant="body-small" className={ styles.error_description }>
+							{ __( 'Check out our troubleshooting guide.', 'jetpack-my-jetpack' ) }
```
This is not a strong opinion, but I think this text may be unnecessary since the CTA says `troubleshoot`. What do you think?

elliottprogrammer replied:
Hmm, well that text _is_ in the Figma design, so I think I should maybe keep it there, at least for now.

![Screen Shot 2024-12-18 at 20 32 57](https://github.com/user-attachments/assets/996fb401-90b6-4adb-a17e-fd5bb081d686)

Figma design: v1IWiVmOAvI8yT8CCa4b1I-fi-4398_7392

CodeyGuyDylan replied:
Is there a copy doc? I've been told the designs aren't always the be-all end-all when it comes to copy 😅

CodeyGuyDylan commented on projects/packages/my-jetpack/_inc/components/product-cards-section/backup-card/index.jsx:
```diff
@@ -128,12 +130,58 @@ const getTimeSinceLastRenewableEvent = lastRewindableEventTime => {
+									status: status,
+									backup_status: lastBackupStatus,
+									feature: 'jetpack-backup',
+									message: '', //errorDescription,
```
Can we remove that commented out variable by chance?

elliottprogrammer replied:
Good catch thanks!

CodeyGuyDylan commented on projects/packages/my-jetpack/_inc/hooks/use-notification-watcher/use-backup-needs-attention-notice.tsx:
```diff
@@ -0,0 +1,128 @@
+ *
+ * @return {Date} a JavaScript Date object
+ */
+function applyTimezone( date: string, offset: number ): Date {
```
This could be a good util function perhaps 🤔

elliottprogrammer replied:
Yes, good point, moved to the utils folder. 👍

CodeyGuyDylan reviewed (dismissed):
Thank you for all your hard work on getting human readable errors! I think it looks great and functions great!

Below is the only blocking change that I found, I left a few other comments but overall this looks great! I am going to throw an approval on this as I don't want to hold up the deployment any longer. If the tooltip issue gets fixed I think this is good to go 😄 But if you'd like me to check it out again after changes just let me know

I am seeing certain screen sizes where the tooltip is off the screen like this:
![image](https://github.com/user-attachments/assets/ef2e45c0-823c-4c2b-b5ab-06dab38b2b5d)

elliottprogrammer committed:
Address various minor feedback.

elliottprogrammer committed:
Merge branch 'trunk' into add/backup-redbubble-needs-attention-notice

elliottprogrammer committed:
When checking for expiring products, dont check Features classes.
